README.md
# mcropper
> 要在移动端实现上传头像，本来采用`cropperjs`，开发完在手机上一跑，哭了。百度之，试用了一下`AlloyCrop`，流畅但是可定制性不强，遂改源码。改着改着就面目全非了，那就尝试发布第一款npm包吧。

## Features
* 只支持移动端
* 裁剪框大小自适应
* 初始化时图片有一边尺寸与裁剪框相等，另一边大于或等于裁剪框且居中
* 缩放图片时，双指中点对应图片的位置会跟随双指中点移动
* 无论移动或缩放，裁剪框始终不会超出图片外
* window.resize时自动重绘，并保持裁剪的区域不变

## Getting started
```js
class MCropper {
  /**
   * @param {Element} container 一个宽度和高度都不为0的可定位元素
   * @param {String} src img src
   * @param {Object} [options]
   * @param {Number} [options.width] 裁剪框宽度(px)
   * @param {Number} [options.height] 裁剪框高度(px)
   * @param {Number} [options.aspectRatio=1] 宽高比，当没有同时指定宽高时起效
   * @param {Number} [options.containRatio=0.92] 当没有指定宽高时，配合aspectRatio自动计算尺寸，以使其宽度和高度适应内容区域，该值控制缩比例(ratio <= 1)
   * @param {Boolean} [options.circle] 是否启用圆形裁剪框，取裁剪框宽度为直径
   * @param {Function(cropper)} [options.onReady] 初始化成功的回调
   * @param {String} [options.borderColor='rgba(51, 153, 255, 0.75)'] 裁剪框边框颜色
   * @param {Number} [options.borderWidth=1] 裁剪框边框宽度(px)
   * @param {Number} [options.borderOrigin=out] 裁剪框边框线绘制位置，out：框的外面，in：框的里面，middle：框的里外各绘制一半
   * @param {Number} [options.modalOpacity=0.6] 蒙层不透明度
   * @param {Boolean} [options.responsive=true] 是否在window.resize后重新渲染
   * @param {Boolean} [options.restore=true] 是否在window.resize重新渲染后保持当前裁剪区域不变
   */
  constructor(container, src, options) {}
  /**
   * 裁剪，下列可选参数有且只有一个发挥作用，由上往下优先级降低
   * @param {Object|Number} [options=1] 为数字时作用和options.ratio一样
   * @param {Number} [options.width] 给出宽度，自动计算高度
   * @param {Number} [options.height] 给出高度，自动计算宽度
   * @param {Number} [options.naturalRatio] 基于图片原始尺寸的输出倍率
   * @param {Number} [options.ratio] 基于裁剪框尺寸的输出倍率
   * @return {Canvas}
   */
  crop(options) {}
  /**
   * 注册初始化成功的回调
   * @param {Function(cropper)} callback 初始化成功后会按顺序执行回调（通过实例化时的options.onReady注册的优先级最高），初始化成功后注册的回调会直接执行
   */
  onReady(callback) {}
  /**
   * 回收资源，移除画布
   */
  destroy() {}
}
```

## Example
```js
new MCropper(container, img.src, {
  onReady(cropper) {
    confirmBtn.onclick = function () {
      preview.src = cropper.crop().toDataURL('image/png')
      confirmBtn.onclick = null
      cropper.destroy()
    }
  }
})
```

## Preview
![Preview](https://dgmpk.github.io/mcropper/public/preview.jpg)

## Demo
[https://dgmpk.github.io/mcropper/](https://dgmpk.github.io/mcropper/)

![QR code](https://dgmpk.github.io/mcropper/public/QR-code.png)

## Dependencies
* [AlloyFinger](https://github.com/AlloyTeam/AlloyFinger)

## License
This content is released under the [MIT](http://opensource.org/licenses/MIT) License.


CHANGELOG.md
# 1.1.0 (2018-10-18)

## Code Refactoring

## Features

* 缩放图片时不再锁定图片的位置，双指中点对应图片的位置会跟随双指中点移动
* 新增可选参数borderOrigin控制裁剪框边框绘制位置
* 现在允许通过修改MCropper.defaults来控制默认值

# 1.1.1 (2019-01-26)

## Bug Fixes

* 修复不传options报错的Bug

# 2.0.0 (2019-08-02)

## Code Refactoring

## Features

* 添加TypeScript支持
* 调整类实例化参数：src变为可选；移除options.onReady并调整为末尾的可选参数；移除options.responsive和options.restore以及各自对应的功能
* 新增实例方法setSrc：实例化可以修改src
* 调整实例方法crop：参数优化
* 调整实例方法resize：新增reset参数，代替类实例化参数options.restore，起相反作用


index.d.ts
export default class MCropper {
  /** 实例化时options参数的默认值，可以通过修改该对象的值影响后续的实例化 */
  static defaults: MCropperOption

  readonly container: string
  readonly src: string

  /**
   * 创建MCropper实例
   * @param {Element} container 一个宽度和高度都不为0的可定位元素
   * @param {string} [src] img src
   * @param {object} [options] 裁剪框配置
   * @param {boolean} [options.circle] 是否启用圆形裁剪框，圆形裁剪框会取裁剪框宽度为直径
   * @param {number} [options.width] 裁剪框宽度(px)
   * @param {number} [options.height] 裁剪框高度(px)
   * @param {number} [options.aspectRatio=1] 裁剪框宽高比，当没有同时指定宽高时起效
   * @param {number} [options.containRatio=0.92] 当裁剪框宽高都没有指定时，配合aspectRatio自动计算尺寸，以使其宽度和高度适应内容区域，该值控制缩比例(ratio <= 1)
   * @param {string} [options.borderColor='rgba(51, 153, 255, 0.75)'] 裁剪框边框颜色
   * @param {number} [options.borderWidth=1] 裁剪框边框宽度(px)
   * @param {number} [options.borderOrigin='out'] 裁剪框边框线绘制位置，out：框的外面，in：框的里面，middle：框的里外各绘制一半
   * @param {number} [options.modalOpacity=0.6] 蒙层不透明度
   * @param {ReadyCallback} [callbackfn] 有src时，该图片加载成功后执行回调
   * @example
   * new MCropper(cropperContainer, '/picture.png', function(cropper) {
   *   cropBtn.addEventListener('click', crop)
   *   function crop () {
   *     previewImg.src = cropper.crop().toDataURL('image/jpeg')
   *     previewImg.style.display = 'block'
   *     cropperContainer.style.display = 'none'
   *     cropBtn.removeEventListener('click', crop)
   *     cropper.destroy()
   *   }
   * })
   */
  constructor(container: Element, src?: string, options?: MCropperOption, callbackfn?: ReadyCallback)
  constructor(container: Element, src: string, callbackfn: ReadyCallback)
  constructor(container: Element, options: MCropperOption)

  /**
   * 裁剪
   * @param {number} [value=1]
   * @param {string} [attribute='ratio'] 指定value的意义：
   *   ratio: 基于裁剪框尺寸的输出倍率；
   *   naturalRatio: 基于图片原始尺寸的输出倍率；
   *   width: 给出宽度，自动计算高度；
   *   height: 给出高度，自动计算宽度；
   * @return {HTMLCanvasElement}
   */
  crop(value?: number, attribute?: 'ratio' | 'naturalRatio' | 'width' | 'height'): HTMLCanvasElement

  /**
   * 更新src（会清空当前回调队列）
   * @param {string} src
   * @param {ReadyCallback} [callbackfn] 图片加载成功后执行回调
   */
  setSrc(src: string, callbackfn?: ReadyCallback): void

  /**
   * 如果当前有已加载成功的图片时马上执行；否则注册至回调队列，在当前图片加载成功后按注册顺序执行
   * @param {ReadyCallback} callbackfn
   */
  onReady(callbackfn: ReadyCallback): void

  /**
   * 在container大小变化后调用以重绘canvas
   * @param {boolean} [reset] 是否在重绘后重置裁剪区域
   */
  resize(reset?: boolean): void

  /**
   * 计算裁剪数据
   * @return {number[]} 前2项：被剪切图像的裁剪起点的 x 坐标和 y 坐标；后2项：被剪切图像的宽度和高度
   */
  getImageData(): [number, number, number, number]

  /** 销毁实例 */
  destroy(): void
}

/**
 * 图片加载完毕后的回调
 * @callback ReadyCallback
 * @param {MCropper} instance
 */
interface ReadyCallback {
  (instance: MCropper): void
}

export interface MCropperOption {
  circle?: boolean
  width?: number
  height?: number
  aspectRatio?: number
  containRatio?: number
  modalOpacity?: number
  borderColor?: string
  borderWidth?: number
  borderOrigin?: 'out' | 'in' | 'middle'
}


.eslintrc.js
// https://eslint.org/docs/user-guide/configuring

module.exports = {
  root: true,
  parser: 'babel-eslint',
  env: {
    browser: true,
  },
  extends: [
    // https://github.com/standard/standard/blob/master/docs/RULES-en.md
    'standard'
  ],
  // add your custom rules here
  rules: {
    // 要求箭头函数的参数使用圆括号
    'arrow-parens': 0,
    // 禁止末尾逗号
    'comma-dangle': 0,
    // allow async-await
    'generator-star-spacing': 0,
    // 强制在关键字前后使用一致的空格
    'keyword-spacing': 0,
    // allow debugger during development
    'no-debugger': process.env.NODE_ENV === 'production' ? 2 : 0,
    // 禁止在使用new构造一个实例后不赋值
    'no-new': 0,
    // 语句强制分号结尾
    'semi': 0,
    // 强制在 function的左括号之前使用一致的空格
    'space-before-function-paren': 0,
    'prefer-promise-reject-errors': 1,
    'standard/no-callback-literal': 0
  }
}


index.js
import AlloyFinger from 'alloyfinger'

const NAMESPACE = '__MCropper__'

const defaults = {
  aspectRatio: 1,
  containRatio: 0.92,
  modalOpacity: 0.6,
  borderColor: 'rgba(51, 153, 255, 0.75)',
  borderWidth: 1,
  borderOrigin: 'out',
}

export default class MCropper {
  constructor(container, src, options, callbackfn) {
    if (container[NAMESPACE]) {
      container[NAMESPACE].destroy()
    }
    container[NAMESPACE] = this
    this.container = container

    if (typeof src === 'object') {
      options = src
      src = null
    } else if (typeof options === 'function') {
      callbackfn = options
      options = null
    }
    this.options = assign({}, MCropper.defaults, options)

    // try to calculate fixed size of crop box if width or height has provided
    const {
      width,
      height,
      aspectRatio
    } = this.options
    if (width || height) {
      if (width) {
        this.cropBoxWidth = width
      }
      if (height) {
        this.cropBoxHeight = height
      }
      if (!this.cropBoxHeight) {
        this.cropBoxHeight = this.cropBoxWidth / aspectRatio
      }
      if (!this.cropBoxWidth) {
        this.cropBoxWidth = this.cropBoxHeight * aspectRatio
      }
    }

    this.canvas = document.createElement('canvas')
    assign(this.canvas.style, {
      position: 'absolute',
      zIndex: '1',
      top: '0',
      left: '0'
    })
    this.ctx = this.canvas.getContext('2d')
    this.drawCropBox = this.options.circle
      ? spread => this.ctx.arc(
        this.container.clientWidth / 2,
        this.container.clientHeight / 2,
        this.cropBoxWidth / 2 + spread / 2,
        0,
        Math.PI * 2,
        false
      )
      : spread => this.ctx.rect(
        this.container.clientWidth / 2 - this.cropBoxWidth / 2 - spread / 2,
        this.container.clientHeight / 2 - this.cropBoxHeight / 2 - spread / 2,
        this.cropBoxWidth + spread,
        this.cropBoxHeight + spread
      )
    this.container.appendChild(this.canvas)

    this.img = new Image()
    src && this.setSrc(src, callbackfn)
  }
  setSrc(src, callbackfn) {
    this.alloyFingerListener && this.alloyFingerListener.destroy()
    this.ready = false
    this.src = src
    this.callbackfns = []
    callbackfn && this.callbackfns.push(callbackfn)
    if (src.substring(0, 4).toLowerCase() === 'http') {
      // resolve base64 uri bug in safari:'cross-origin image load denied by cross-origin resource sharing policy.'
      this.img.crossOrigin = 'anonymous'
    }
    this.img.onload = this.init.bind(this)
    this.img.src = src
  }
  init() {
    this.ready = true
    this.calculateImgData()
    this.renderCenter()
    let callback
    while ((callback = this.callbackfns.shift())) {
      callback(this)
    }
    this.addFingerListener()
  }
  calculateImgData() {
    const containerWidth = this.canvas.width = this.container.clientWidth
    const containerHeight = this.canvas.height = this.container.clientHeight
    const {
      width,
      height,
      aspectRatio,
      containRatio,
    } = this.options
    if (!width && !height) {
      if (aspectRatio > containerWidth / containerHeight) {
        this.cropBoxWidth = containerWidth * containRatio
        this.cropBoxHeight = this.cropBoxWidth / aspectRatio
      } else {
        this.cropBoxHeight = containerHeight * containRatio
        this.cropBoxWidth = this.cropBoxHeight * aspectRatio
      }
    }
    this.cropBoxOriginX = (containerWidth - this.cropBoxWidth) / 2
    this.cropBoxOriginY = (containerHeight - this.cropBoxHeight) / 2
    this.imgMinScale = Math.max(
      this.cropBoxWidth / this.img.naturalWidth,
      this.cropBoxHeight / this.img.naturalHeight
    )
  }
  renderCenter() {
    const imgScale = this.imgMinScale
    this.render(
      this.cropBoxOriginX - (this.img.naturalWidth * imgScale - this.cropBoxWidth) / 2,
      this.cropBoxOriginY - (this.img.naturalHeight * imgScale - this.cropBoxHeight) / 2,
      imgScale
    )
  }
  render(imgOriginX, imgOriginY, imgScale) {
    this.renderCover()
    this.renderImage(imgOriginX, imgOriginY, imgScale)
  }
  renderCover() {
    const {
      ctx,
      options: {
        modalOpacity,
        borderColor,
        borderWidth,
        borderOrigin
      }
    } = this

    // renderShadow
    ctx.save()
    ctx.globalCompositeOperation = 'copy' // clear old
    ctx.globalAlpha = modalOpacity
    ctx.fillStyle = 'black'
    ctx.fillRect(0, 0, this.container.clientWidth, this.container.clientHeight)
    ctx.restore()

    const spread = borderOrigin === 'out'
      ? borderWidth
      : borderOrigin === 'in'
        ? -borderWidth
        : 0

    // strokeCropBox
    ctx.save()
    ctx.beginPath()
    ctx.globalCompositeOperation = 'destination-out' // 裁剪阴影的一部分
    this.drawCropBox(spread + borderWidth) // 加大裁剪部分的面积，使边框线不会重叠在阴影上方
    ctx.fill()
    ctx.restore()

    // renderCropBoxBorder
    ctx.save()
    ctx.beginPath()
    ctx.strokeStyle = borderColor
    ctx.lineWidth = borderWidth
    this.drawCropBox(spread)
    ctx.stroke()
    ctx.restore()
  }
  renderImage(imgOriginX, imgOriginY, imgScale) {
    imgScale = Math.max(imgScale, this.imgMinScale)

    // 根据缩放计算一个img起点坐标范围来限制img起点坐标的取值范围，以保证图片始终比裁剪框大
    if (imgScale !== this.imgScale) {
      this.imgScale = imgScale
      this.imgWidth = this.img.naturalWidth * imgScale
      this.imgHeight = this.img.naturalHeight * imgScale
      const imgMaxOffsetX = this.imgWidth - this.cropBoxWidth
      const imgMaxOffsetY = this.imgHeight - this.cropBoxHeight
      this.imgOriginXRange = [this.cropBoxOriginX - imgMaxOffsetX, this.cropBoxOriginX]
      this.imgOriginYRange = [this.cropBoxOriginY - imgMaxOffsetY, this.cropBoxOriginY]
    }
    this.imgOriginX = limitRange(imgOriginX, this.imgOriginXRange)
    this.imgOriginY = limitRange(imgOriginY, this.imgOriginYRange)

    const ctx = this.ctx
    ctx.save()
    ctx.globalCompositeOperation = 'destination-over'
    ctx.drawImage(
      this.img,
      this.imgOriginX,
      this.imgOriginY,
      this.imgWidth,
      this.imgHeight
    )
    ctx.restore()
  }

  /**
   * 缩放（始终以多指触点的中点缩放，所以可能会同时带有平移）
   * @param {number} scale
   * @param {[number, number]} param1 缩放前的中点在image坐标系的坐标
   * @param {[number, number]} param2 缩放后的中点在container坐标系的坐标
   */
  zoomImage(scale, [ix, iy], [cx, cy]) {
    scale = Math.max(this.imgMinScale, scale)
    const zoom = scale / this.imgScale
    this.render(cx - ix * zoom, cy - iy * zoom, scale)
  }
  pagePointToContainerPoint([pageX, pageY]) {
    const {
      top,
      left
    } = this.container.getBoundingClientRect()
    return [pageX - left, pageY - top]
  }
  containerPointToImagePoint([cx, cy]) {
    return [cx - this.imgOriginX, cy - this.imgOriginY]
  }
  addFingerListener() {
    let initialImgScale
    let centerOfGravityOnImage
    const getCenterOfGravityOnContainer = event => {
      const points = [].slice.apply(event.touches).map(touch => [touch.pageX, touch.pageY])
      let centerOfGravity
      if (points.length === 1) {
        centerOfGravity = points[0]
      } else if (points.length === 2) {
        centerOfGravity = [(points[0][0] + points[1][0]) / 2, (points[0][1] + points[1][1]) / 2]
      } else {
        centerOfGravity = getCenterOfGravity(points)
      }
      return this.pagePointToContainerPoint(centerOfGravity)
    }
    const updateCenterOfGravityOnImage = containerPoint => {
      centerOfGravityOnImage = this.containerPointToImagePoint(containerPoint)
    }
    this.alloyFingerListener = new AlloyFinger(this.canvas, {
      multipointStart: e => {
        initialImgScale = this.imgScale
        updateCenterOfGravityOnImage(getCenterOfGravityOnContainer(e))
      },
      pinch: e => {
        const newCenterOfGravityOnContainer = getCenterOfGravityOnContainer(e)
        this.zoomImage(initialImgScale * e.zoom, centerOfGravityOnImage, newCenterOfGravityOnContainer)
        updateCenterOfGravityOnImage(newCenterOfGravityOnContainer)
      },
      pressMove: e => {
        e.preventDefault()
        this.render(this.imgOriginX + e.deltaX, this.imgOriginY + e.deltaY, this.imgScale)
      }
    })
  }
  resize(reset) {
    if (reset) {
      this.calculateImgData()
      this.renderCenter()
    } else {
      const [
        sx,
        sy
      ] = this.getImageData()
      const zoom = this.imgScale / this.imgMinScale
      this.calculateImgData()
      const imgScale = this.imgMinScale * zoom
      this.render(
        this.cropBoxOriginX - sx * imgScale,
        this.cropBoxOriginY - sy * imgScale,
        imgScale
      )
    }
  }
  getImageData() {
    const scale = this.imgScale
    return [
      (this.cropBoxOriginX - this.imgOriginX) / scale, // 被剪切图像的裁剪起点的 x 坐标
      (this.cropBoxOriginY - this.imgOriginY) / scale, // 被剪切图像的裁剪起点的 y 坐标
      this.cropBoxWidth / scale, // 被剪切图像的宽度
      this.cropBoxHeight / scale // 被剪切图像的高度
    ]
  }
  crop(value = 1, attribute = 'ratio') {
    if (!this.ready) {
      throw new Error('can\'t crop before ready')
    }
    const canvas = document.createElement('canvas')
    const ctx = canvas.getContext('2d')
    const aspectRatio = this.cropBoxWidth / this.cropBoxHeight
    const imageData = this.getImageData()
    switch (attribute) {
      case 'width':
        canvas.width = value
        canvas.height = value * aspectRatio
        break;
      case 'height':
        canvas.height = value
        canvas.width = value / aspectRatio
        break;
      case 'naturalRatio':
        canvas.width = imageData[2] * value
        canvas.height = imageData[3] * value
        break;
      default:
        canvas.width = this.cropBoxWidth * value
        canvas.height = this.cropBoxHeight * value
    }
    ctx.drawImage(
      this.img,
      imageData[0],
      imageData[1],
      imageData[2],
      imageData[3],
      0,
      0,
      canvas.width,
      canvas.height
    )
    return canvas
  }
  onReady(callback) {
    if (this.ready === true) {
      callback(this)
    } else {
      this.callbackfns.push(callback)
    }
  }
  destroy() {
    this.img.src = this.img.onload = null
    this.alloyFingerListener && this.alloyFingerListener.destroy()
    this.container.removeChild(this.canvas)
    delete this.container[NAMESPACE]
  }
}

function limitRange(val, [min, max]) {
  return val < min ? min : val > max ? max : val
}

function assign(object, ...sources) {
  sources.forEach(source => {
    for (const variable in source) {
      if (source.hasOwnProperty(variable)) {
        object[variable] = source[variable]
      }
    }
  })
  return object
}

function getCenterOfGravity(points) {
  const len = points.length
  let area = 0
  let x = 0
  let y = 0

  for (let index = 1; index <= len; index++) {
    const [nx, ny] = points[index % len]
    const [lx, ly] = points[index - 1]
    const temp = (nx * ly - ny * lx) / 2
    area += temp
    x += temp * (nx + lx) / 3
    y += temp * (ny + ly) / 3
  }
  x = x / area
  y = y / area
  return [x, y]
}

MCropper.defaults = defaults
